File: leetcode/Algorithms/110.BalancedBinaryTree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def isBalanced(self, root):
        """
        :type root: TreeNode
        :rtype: bool
        """
        return self.dfsHeight(root) != -1

    def dfsHeight(self, root):
        if not root:
            return 0

        left_height = self.dfsHeight(root.left)
        right_height = self.dfsHeight(root.right)
        if left_height == -1 or right_height == -1:
            return -1
        if abs(left_height - right_height) > 1:
            return -1

        return max(left_height, right_height) + 1

File: leetcode/Algorithms/110.BalancedBinaryTree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

public class Solution {
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;
        return Math.abs(treeDepth(root.right) - treeDepth(root.left)) <= 1
                && isBalanced(root.left)
                && isBalanced(root.right);
    }

    private int treeDepth(TreeNode root) {
        return root == null ? 0 : Math.max(treeDepth(root.left), treeDepth(root.right)) + 1;
    }
}
